packages/combobox/src/Combobox.tsx

```diff
@@ -0,0 +1,923 @@
+      const safeSpace = Math.max(
+        viewportSize.height - triggerBottom,
+        triggerTop,
+      );
+
+      // if there's more than enough space, set to maxMenuHeight
+      // otherwise fill the space available
+      return Math.min(maxMenuHeight, safeSpace - menuMargin);
+    }
+
+    return maxMenuHeight;
+  }, [viewportSize, comboboxRef, menuRef]);
+
+  // Scroll the menu when the focus changes
+  useEffect(() => {
+    // get the focused option
+  }, [focusedOption]);
+
+  /**
+   *
+   * Event Handlers
+   *
+   */
+  // Fired when the wrapper gains focus
+  const handleInputFocus = () => {
+    scrollToEnd();
+    openMenu();
+  };
+
+  // Fired onChange
+  const handleInputChange = ({
+    target: { value },
+  }: React.ChangeEvent<HTMLInputElement>) => {
+    setInputValue(value);
+    // fire any filter function passed in
+    onFilter?.(value);
+  };
+
+  const handleClearButtonFocus = () => {
+    setFocusedOption(null);
+  };
+
+  const handleKeyDown = (event: React.KeyboardEvent) => {
+    const isFocusInMenu = menuRef.current?.contains(document.activeElement);
+    const isFocusOnCombobox = comboboxRef.current?.contains(
+      document.activeElement,
+    );
+
+    const isFocusInComponent = isFocusOnCombobox || isFocusInMenu;
+
+    if (isFocusInComponent) {
+      // No support for modifiers yet
+      if (event.ctrlKey || event.shiftKey || event.altKey) {
+        return;
+      }
+
+      switch (event.keyCode) {
+        case keyMap.Tab:
+        case keyMap.Escape: {
+          closeMenu();
+          updateFocusedOption('first');
+          break;
+        }
+
+        case keyMap.Enter:
+        case keyMap.Space: {
+          if (isOpen) {
+            // prevent typing the space character
+            event.preventDefault();
+          }
+
+          if (
+            // Focused on input element
+            document.activeElement === inputRef.current &&
+            isOpen &&
+            !isNull(focusedOption)
+          ) {
+            updateSelection(focusedOption);
+          } else if (
+            // Focused on clear button
+            document.activeElement === clearButtonRef.current
+          ) {
+            updateSelection(null);
+            setInputFocus();
+          }
+          break;
+        }
+
+        case keyMap.ArrowDown: {
+          if (isOpen) {
+            // Prevent the page from scrolling
+            event.preventDefault();
+          }
+          openMenu();
+          updateFocusedOption('next');
+          break;
+        }
+
+        case keyMap.ArrowUp: {
+          if (isOpen) {
+            // Prevent the page from scrolling
+            event.preventDefault();
+          }
+          updateFocusedOption('prev');
+          break;
+        }
+
+        case keyMap.ArrowRight: {
+          updateFocusedChip('next');
+          break;
+        }
+
+        case keyMap.ArrowLeft: {
+          updateFocusedChip('prev');
```

Comment:
Currently, this handling seems to break the ability to use arrow keys to move through a string that you've typed. It'll be a challenge, but we need to only move focus to a chip from the input when your cursor is all the way to the left.

I think there's a DOM API like `selectionStart` you might be able to use for this.